pandas/core/reshape/merge.py

```diff
@@ -1730,6 +1730,9 @@ def flip(xs) -> np.ndarray:
         )
         tolerance = self.tolerance
 
+        if is_object_dtype(right_values.dtype) or is_object_dtype(left_values.dtype):
```

Comment:
see if you can integrate this into _get_merge_keys; its in a pretty random place here.

Reply:
I think we actually need to systematically check the merge operands more, like we do for other types of merging. 

We also generally raise ValueError and not TypeError

Reply:
I had ValueError initially :) I will change it back

Reply:
@jreback 
I moved the test. Is this what you had in mind?

Reply:
hmm, we actually need to check if the left/right are the *same* type. I *think* we already do this, can you please try to trace the code a bit more, e.g. pass datetime on one side and integer on the other.

Reply:
yeah you are right. We do this in  _get_merge_keys. But object dtype is allowed there , because the by key can be of object dtype. 

It is in https://github.com/pandas-dev/pandas/blob/master/pandas/core/reshape/merge.py#1643

If we want to check the object dtype there, we have to check if this is not part of the by parameter and perform the check afterwards. I thought this would be a bit confusing.

Reply:
no you don't, simply override _get_merge_keys and add this first, then call super(). this is allowed for non-merge_asof but not merge_asof.

also ideally expand the tests for disallowed indices (e.g. the example i gave should *already* fail).

Reply:
Your example already fails, tests exist already.

I'm sorry, if I am a bit slow here, but the function I referenced already overwrites the original _get_merge_key function. It calls super at first and checks afterwards for equal dtypes. And then it returns the left join keys and right join keys which are the union of the by column and the on column in case of merge_asof.